shell/browser/api/electron_api_utility_process.cc

```diff
@@ -0,0 +1,642 @@
+                       std::move(message)));
+  }
+
+  void ShutdownReader() override {
+    content::GetUIThreadTaskRunner({})->PostTask(
+        FROM_HERE,
+        base::BindOnce(&api::UtilityProcessWrapper::ShutdownStdioReader,
+                       utility_process_wrapper_,
+                       api::UtilityProcessWrapper::IOHandle::STDOUT));
+  }
+
+  base::WeakPtr<api::UtilityProcessWrapper> utility_process_wrapper_;
+};
+
+class StderrPipeReader : public PipeReaderBase {
+ public:
+  StderrPipeReader(
+      base::WeakPtr<api::UtilityProcessWrapper> utility_process_wrapper,
+#if BUILDFLAG(IS_WIN)
+      HANDLE read_handle)
+      : PipeReaderBase("UtilityProcessStderrReadThread", read_handle),
+#else
+      int read_fd)
+      : PipeReaderBase("UtilityProcessStderrReadThread", read_fd),
+#endif
+        utility_process_wrapper_(std::move(utility_process_wrapper)) {
+  }
+
+ private:
+  void HandleMessage(std::vector<uint8_t> message) override {
+    content::GetUIThreadTaskRunner({})->PostTask(
+        FROM_HERE,
+        base::BindOnce(&api::UtilityProcessWrapper::HandleStdioMessage,
+                       utility_process_wrapper_,
+                       api::UtilityProcessWrapper::IOHandle::STDERR,
+                       std::move(message)));
+  }
+
+  void ShutdownReader() override {
+    content::GetUIThreadTaskRunner({})->PostTask(
+        FROM_HERE,
+        base::BindOnce(&api::UtilityProcessWrapper::ShutdownStdioReader,
+                       utility_process_wrapper_,
+                       api::UtilityProcessWrapper::IOHandle::STDERR));
+  }
+
+  base::WeakPtr<api::UtilityProcessWrapper> utility_process_wrapper_;
+};
+
+namespace api {
+
+gin::WrapperInfo UtilityProcessWrapper::kWrapperInfo = {
+    gin::kEmbedderNativeGin};
+
+UtilityProcessWrapper::UtilityProcessWrapper(
+    node::mojom::NodeServiceParamsPtr params,
+    std::u16string display_name,
+    std::map<IOHandle, IOType> stdio,
+    base::EnvironmentMap env_map,
+    base::FilePath current_working_directory,
+    bool use_plugin_helper) {
+#if BUILDFLAG(IS_WIN)
+  base::win::ScopedHandle stdout_write(nullptr);
+  base::win::ScopedHandle stderr_write(nullptr);
+#elif BUILDFLAG(IS_POSIX)
+  base::FileHandleMappingVector fds_to_remap;
+#endif
+  for (const auto& [io_handle, io_type] : stdio) {
+    if (io_type == IOType::PIPE) {
+#if BUILDFLAG(IS_WIN)
+      HANDLE read = nullptr;
+      HANDLE write = nullptr;
+      if (!::CreatePipe(&read, &write, nullptr, 0)) {
+        PLOG(ERROR) << "pipe creation failed";
+        return;
+      }
+      if (io_handle == IOHandle::STDOUT) {
+        stdout_write.Set(write);
+        stdout_reader_ = std::make_unique<StdoutPipeReader>(
+            weak_factory_.GetWeakPtr(), read);
+        if (!stdout_reader_->Start()) {
+          PLOG(ERROR) << "stdout output watcher thread failed to start.";
+          PipeIOBase::Shutdown(std::move(stdout_reader_));
+          CloseHandle(write);
```

Comment:
Is this necessary? `stdout_write` already manages the `write` handle.

Reply:
Yeah not necessary to do an explicit close.